notify/factory.go

```diff
@@ -27,57 +31,147 @@ import (
-	factory, exists := receiverFactories[receiverType]
-	return factory, exists
-}
-
-// wrap wraps the notifier's factory errors with receivers.ReceiverValidationError
-func wrap[T NotificationChannel](f func(fc receivers.FactoryConfig) (T, error)) func(receivers.FactoryConfig) (NotificationChannel, error) {
-	return func(fc receivers.FactoryConfig) (NotificationChannel, error) {
-		ch, err := f(fc)
-		if err != nil {
-			return nil, ReceiverValidationError{
-				Err: err,
-				// TODO it will be removed in the next PR
-				Cfg: &GrafanaReceiver{
-					UID:                   fc.Config.UID,
-					Name:                  fc.Config.Name,
-					Type:                  fc.Config.Type,
-					DisableResolveMessage: fc.Config.DisableResolveMessage,
-					Settings:              fc.Config.Settings,
-					SecureSettings:        nil,
-				},
+// BuildReceiverIntegrations creates integrations for each configured notification channel in GrafanaReceiverConfig.
+// It returns a slice of Integration objects, one for each notification channel, along with any errors that occurred.
+func BuildReceiverIntegrations(
+	receiver GrafanaReceiverConfig,
+	tmpl *template.Template,
+	img images.ImageStore, // Used by some receivers to include as part of the source
+	newLogger logging.LoggerFactory,
+	newWebhookSender func(n receivers.Metadata) (receivers.WebhookSender, error),
+	newEmailSender func(n receivers.Metadata) (receivers.EmailSender, error),
+	orgID int64,
+	version string,
+) ([]*Integration, error) {
+	type notificationChannel interface {
+		notify.Notifier
+		notify.ResolvedSender
+	}
+	var (
+		integrations []*Integration
+		errors       types.MultiError
+		// Helper function to create an integration for a notification channel and add it to the integrations slice.
+		createIntegration = func(idx int, cfg receivers.Metadata, f func(logger logging.Logger) notificationChannel) {
+			logger := newLogger("ngalert.notifier."+cfg.Type, "notifierUID", cfg.UID)
+			n := f(logger)
+			i := NewIntegration(n, n, cfg.Type, idx)
+			integrations = append(integrations, i)
+		}
+		// Helper function to create an integration for a notification channel that requires a webhook sender.
+		createIntegrationWithWebhook = func(idx int, cfg receivers.Metadata, f func(logger logging.Logger, w receivers.WebhookSender) notificationChannel) {
+			w, e := newWebhookSender(cfg)
+			if e != nil {
+				errors.Add(fmt.Errorf("unable to build webhook client for %s notifier %s (UID: %s): %w ", cfg.Type, cfg.Name, cfg.UID, e))
+				return
 			}
+			createIntegration(idx, cfg, func(logger logging.Logger) notificationChannel {
+				return f(logger, w)
+			})
 		}
-		return ch, nil
+	)
+	// Range through each notification channel in the receiver and create an integration for it.
+	for i, cfg := range receiver.AlertmanagerConfigs {
+		createIntegration(i, cfg.Metadata, func(l logging.Logger) notificationChannel {
+			return alertmanager.New(cfg.Settings, cfg.Metadata, img, l)
+		})
```

Comment:
nit: As we discussed offline there is another approach here to take, you can use more in-line function to try and simplify this createIntegration - it wasn't super clear why this was here.

```golang
		createIntegration(i, cfg.Metadata, dinding.New(cfg.Settings, cfg.Metadata, tmpl, nw(cfg.Metadata), nl(logger)))
```